Use flake8 directly

nuclearsandwich opened the pull request:
This PR breaks the repository level dependency cycle of `ament_package` --test-depend--> `ament_lint` --buildtool_depend--> `ament_cmake_core` --buildtool_depend--> `ament_package` by removing the pep257 and copyright checks and using flake8 directly for the flake8 checks.

This also means losing the JUnit xml output from ament_flake8 for this package.

On this initial review pass I'd like to know if there's any additional functionality we should add before CI and final review.

A complete alternative would be skipping pytest entirely and running flake8 directly as the test command, but I'm not completely certain of the details or implications of doing so.

dirk-thomas commented on test/test_flake8.py:
```diff
@@ -11,13 +11,21 @@
+    # Configuration options adapted from the ament_flake8
+    # https://github.com/ament/ament_lint/blob/master/ament_flake8/ament_flake8/configuration/ament_flake8.ini
+    style = get_style_guide(
+        ignore='C816,D100,D101,D102,D103,D104,D105,D106,D107,D203,D212,D404,I202',
```
Not all of these are probably necessary. I would suggest to remove the ones not relevant for this package.

nuclearsandwich replied:
In order to do that there needs to be a shared understanding of what error codes are relevant.

I can think of one heuristic which would be with all flake8 plugins from CI installed: Do any ignored checks trigger? The ones that do should stay and others can be removed.

Absent doing the above or some other heuristic test, I would need some criteria to decide which are applicable and audit each individual ignored rule.

dirk-thomas replied:
Currently CI passes for this packages. After the patch that should be the same - without modifying any code and without suppressing warning classes which are not even emitted.

nuclearsandwich replied:
Okay so that's a preference for only ignoring warnings that are suppressed in the current codebase rather than suppressing warnings consistently across this package and packages that use ament_lint. :+1:

dirk-thomas commented on test/test_flake8.py:
```diff
@@ -11,13 +11,21 @@
-from ament_flake8.main import main
 import pytest
 
+from flake8.api.legacy import get_style_guide
```
Why is the new import in a different location than the previous one? Based on the naming I would expect it to be the same.

dirk-thomas commented on setup.py:
```diff
@@ -24,7 +24,7 @@
 and provides tooling to build these federated packages together.""",
     license='Apache License, Version 2.0',
-    tests_require=['pytest'],
+    tests_require=['pytest', 'flake8'],
```
Nitpick: alphabetical order.

dirk-thomas commented on test/test_flake8.py:
```diff
@@ -11,13 +11,21 @@
+    # https://github.com/ament/ament_lint/blob/master/ament_flake8/ament_flake8/configuration/ament_flake8.ini
+    style = get_style_guide(
+        ignore='C816,D100,D101,D102,D103,D104,D105,D106,D107,D203,D212,D404,I202',
+        import_order_style='google',
```
Maybe this option isn't being applied and that is why the import above changes in style?

nuclearsandwich replied:
This option relies on a flake8 plugin `flake8-import-order` which wasn't present in my virtualenv. Once the plugin was installed the import order issue above was reported correctly.

There's no apt package for the flake8 plugin which means there's not currently a way to introduce this in a way that would be bloom-able.

dirk-thomas replied:
Actually `flake8-import-error` should probably be listed in `tests_require` in the `setup.py` for it to work correctly in any environment.

nuclearsandwich commented:
Thanks for the review. I'll work on the changes.

> Please run CI for this package.

Definitely planning to run CI before merging any changes here. Goal for this round one was to make sure I am on the right track wrt running flake8 via pytest. Not much sense in tying up CI to test the wrong thing. Once I've addressed the comments above I'll trigger a CI run.

nuclearsandwich commented:
CI on the full complement of ignores.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4641)](http://ci.ros2.org/job/ci_linux/4641/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1538)](http://ci.ros2.org/job/ci_linux-aarch64/1538/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3837)](http://ci.ros2.org/job/ci_osx/3837/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4718)](http://ci.ros2.org/job/ci_windows/4718/)

mikaelarguedas commented:
> This failure is related to how supplemental repos files work. It's on my wannado list for post-bouncy.

While this is a nice to have for external contributions, the development team should really try to use branches on the actual repo rather than forks so that out of date branches are merged with the current master on CI

dirk-thomas commented on test/test_flake8.py:
```diff
@@ -11,13 +11,19 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from flake8.api.legacy import get_style_guide
```
The empty lines were shuffled here. There should be one empty line between the copyright and the first import and no empty lines between the imports (since they are both in the same group).

dirk-thomas commented on test/test_flake8.py:
```diff
@@ -11,13 +11,19 @@
-    assert rc == 0, 'Found code style errors / warnings'
+    style = get_style_guide(
+        ignore='I202',
+        import_order_style='google',
```
As @mikaelarguedas mentioned I don't think this option is currently being applied.

nuclearsandwich replied:
This seems to be the case. Without flake8-import-order installed flake8 logs an error when this option is used. With flake8-import-order installed it does not, but doesn't seem to report errors with import order when invoked via test_flake8.py.

Should we switch to a configuration file instead?

dirk-thomas replied:
Well, currently the code fails the test (at least for me). And changing the argument for the import order style in the test doesn't make any difference.

Only with a `.flake8` file with the following content it does pass for me:

```
[flake8]
import-order-style = google
```

mikaelarguedas commented on test/test_flake8.py:
```diff
@@ -12,12 +12,18 @@
+        statistics=True)
+    results = style.check_files()
+    assert results.total_errors == 0, 'Found code style errors / warnings'
+    assert 1 == 0
```
Debug statement commited by mistake ?

nuclearsandwich replied:
YES. Ugh I should not be up this late.

nuclearsandwich commented:
I guess I'm glad that CI exists because no matter what I can't seem to get my locally run flake8 to behave the same as CI.

nuclearsandwich commented:
> I guess I'm glad that CI exists because no matter what I can't seem to get my locally run flake8 to behave the same as CI.

My local environment had a flake8 config also based on ament_flake8's that was interfering with local testing. Nulling that out I can reproduce CI changes locally.

The current state of this branch still fails with I201 due to the import order style config being ignored.

From https://github.com/ros-infrastructure/catkin_pkg/pull/219#issuecomment-396092843 it looks like the only methods to use a config file rather than the flake8 legacy API are to shell out to a flake8 process in the test or copy over the use of the non-public API used by ament_flake8.

dirk-thomas commented on test/test_flake8.py:
```diff
@@ -12,12 +12,20 @@
+        ignore='D100,D103,D104',
+        # flake8 doesn't appear to honor the import_order_style directive via this api.
+        # https://github.com/ros-infrastructure/catkin_pkg/pull/219#issuecomment-396092843
+        import_order_style='google',
```
Can the option as well as the comment be removed since it doesn't have any effect anyway?

dirk-thomas commented:
> From ros-infrastructure/catkin_pkg#219 (comment) it looks like the only methods to use a config file rather than the flake8 legacy API are to shell out to a flake8 process in the test or copy over the use of the non-public API used by ament_flake8.

Can you please describe in which cases the test currently fails and what you have tried so far?

nuclearsandwich commented:
> Can you please describe in which cases the test currently fails and what you have tried so far?

Currently the import check I201 fails since the wrong import order style is used.

* Running python -mpytest locally with no flake8 config file. I had to temporarily move ~/.config/flake8
```
 ❯ python -mpip list
Package               Version
--------------------- -------
ament-package         0.4.0  
atomicwrites          1.1.5  
attrs                 18.1.0 
flake8                3.5.0  
flake8-blind-except   0.1.1  
flake8-builtins       1.4.1  
flake8-class-newline  1.6.0  
flake8-comprehensions 1.4.1  
flake8-deprecated     1.3    
flake8-docstrings     1.3.0  
flake8-import-order   0.17.1 
flake8-polyfill       1.0.2  
flake8-quotes         1.0.0  
mccabe                0.6.1  
more-itertools        4.2.0  
pip                   10.0.1 
pkg-resources         0.0.0  
pluggy                0.6.0  
py                    1.5.3  
pycodestyle           2.3.1  
pydocstyle            2.1.1  
pyflakes              1.6.0  
pytest                3.6.1  
setuptools            39.2.0 
six                   1.11.0 
snowballstemmer       1.2.1  

 ❯ python -mpytest
... some output elided ...
./test/test_flake8.py:16:1: I201 Missing newline between import groups. 'import pytest' is identified as Third Party and 'from flake8.api.legacy import get_style_guide' is identified as Third Party.
import pytest
^
```

* Running pytest via CI: https://ci.ros2.org/view/nightly/job/ci_linux/4643/console#console-section-6 The D10* failures from this CI run are now ignored again.

I have tried setting the option via `get_style_guide` option. Which doesn't work (see https://github.com/ros-infrastructure/catkin_pkg/pull/219) and I've tried setting `import-order-style` via `~/.config/flake8` which does work but I'm not sure how to introduce a specific config file using the legacy API.

mikaelarguedas commented:
> Late breaking edit: putting configuration in .flake8 in the ament_package repository root works.

I think that's what we should do for this package. It's a bit of a shame to duplicate the configuration but it's better to be consistent and seems to be the simplest way of getting flake8 to honor the option

dirk-thomas commented:
The goal is that this needs to pass testing when `flake8-import-order` is not installed (as described by the package dependencies since it doesn't existing from Debian package) as well as if it is installed (e.g. on a developer machine).

In general it is recommended that developers should not have custom global config files to not interfere.

My earlier suggested approach creating the `.flake8` file is completely sufficient to achieve this (according to my testing). There is also no "duplication" necessary with that approach. Is there any reason that is not the case why something else would be necessary?

mikaelarguedas commented:
>  Is there any reason that is not the case why something else would be necessary?

Not that I know of

nuclearsandwich commented:
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4651)](https://ci.ros2.org/job/ci_linux/4651/)

CI job that just runs the ament_package build and test to verify that it passes CI (where the extra flake8 plugins are installed.

nuclearsandwich commented:
> Can you please describe in which cases the test currently fails and what you have tried so far?

I found this question really helpful by the way. I've got a habit of jumping to the "what next?" when the "what now?" is more useful to reviewers. Thanks!
